casbin/enforcer_synced.cpp

```diff
@@ -0,0 +1,505 @@
+ * @param policyFile the path of the policy file.
+ */
+SyncedEnforcer ::SyncedEnforcer(string model_path, string policy_file)
+  : Enforcer(model_path, policy_file), autoLoadRunning(false)
+{}
+
+/**
+ * Enforcer initializes an enforcer with a database adapter.
+ *
+ * @param model_path the path of the model file.
+ * @param adapter the adapter.
+ */
+SyncedEnforcer ::SyncedEnforcer(string model_path, shared_ptr<Adapter> adapter)
+  : Enforcer(model_path,adapter), autoLoadRunning(false)
+{}
+
+/**
+ * Enforcer initializes an enforcer with a model and a database adapter.
+ *
+ * @param m the model.
+ * @param adapter the adapter.
+ */
+SyncedEnforcer :: SyncedEnforcer(shared_ptr<Model> m, shared_ptr<Adapter> adapter)
+  : Enforcer(m, adapter), autoLoadRunning(false)
+{}
+
+/**
+ * Enforcer initializes an enforcer with a model.
+ *
+ * @param m the model.
+ */
+SyncedEnforcer ::SyncedEnforcer(shared_ptr<Model> m)
+  : Enforcer(m), autoLoadRunning(false)
+{}
+
+/**
+ * Enforcer initializes an enforcer with a model file.
+ *
+ * @param model_path the path of the model file.
+ */
+SyncedEnforcer ::SyncedEnforcer(string model_path)
+  : Enforcer(model_path), autoLoadRunning(false)
+{}
+
+/**
+ * Enforcer initializes an enforcer with a model file, a policy file and an enable log flag.
+ *
+ * @param model_path the path of the model file.
+ * @param policyFile the path of the policy file.
+ * @param enableLog whether to enable Casbin's log.
+ */
+SyncedEnforcer :: SyncedEnforcer(string model_path, string policy_file, bool enable_log)
+  : Enforcer(model_path, policy_file, enable_log), autoLoadRunning(false) 
+{}
+
+// SyncedEnforcer::SyncedEnforcer(const SyncedEnforcer& ce)
+//    : Enforcer(ce), autoLoadRunning(ce.autoLoadRunning)
+// {}
+
+// SyncedEnforcer::SyncedEnforcer(SyncedEnforcer&& ce)
+//   : Enforcer(ce), autoLoadRunning(ce->autoLoadRunning)
+// {}
+
+
+void SyncedEnforcer ::LoadPolicyWrapper() {
+  Enforcer::LoadPolicy();
+  ++n;
+}
+
+// StartAutoLoadPolicy starts a thread that will go through every specified duration call LoadPolicy
+void SyncedEnforcer ::StartAutoLoadPolicy(chrono::duration<int64_t, std::nano> t) {
+  if(IsAutoLoadingRunning())
+    return;
+  autoLoadRunning = true;
+  function<void()> onTick = [this]() {
+    Enforcer::LoadPolicy();
+    ++n;
+  };
+  ticker = unique_ptr<Ticker>(new Ticker(onTick, t));
+  n = 1;
+  ticker->start();
+}
+
+// IsAutoLoadingRunning check if SyncedEnforcer is auto loading policies
+inline bool SyncedEnforcer ::IsAutoLoadingRunning() {
+  return autoLoadRunning;
+}
+
+// StopAutoLoadPolicy causes the thread to exit
+void SyncedEnforcer ::StopAutoLoadPolicy() {
+  ticker->stop();
+  autoLoadRunning = false;
+}
+
+string SyncedEnforcer ::UpdateWrapper() {
+  LoadPolicy();
+  return "";
+}
+
+// SetWatcher sets the current watcher.
+void SyncedEnforcer ::SetWatcher(shared_ptr<Watcher> w){
+  watcher = w;
+  return watcher->SetUpdateCallback(&SyncedEnforcer::UpdateWrapper);
```

Comment:
@xcaptain I think this would be of some context to https://github.com/casbin/casbin-cpp/issues/38#issuecomment-800147824.

Reply:
@EmperorYP7 Nice, and I think you can also add `SetWatcher` to `enforcer`

Reply:
Sure thing!